subprojects/workers/src/main/java/org/gradle/workers/internal/InProcessCompilerDaemonFactory.java

```diff
@@ -0,0 +1,144 @@
+                BuildOperationWorkerRegistry.Completion workerLease = parentWorkerOperation.operationStart();
+                BuildOperationDetails buildOperation = BuildOperationDetails.displayName(action.getDescription()).parent(parentBuildOperation).build();
+                try {
+                    return buildOperationExecutor.run(buildOperation, new Transformer<DefaultWorkResult, BuildOperationContext>() {
+                        @Override
+                        public DefaultWorkResult transform(BuildOperationContext buildOperationContext) {
+                            return executeInWorkerClassLoader(serverImplementationClass, action, spec, forkOptions.getClasspath(), forkOptions.getSharedPackages());
+                        }
+                    });
+                } finally {
+                    workerLease.operationFinish();
+                }
+            }
+        };
+    }
+
+    private <T extends WorkSpec> DefaultWorkResult executeInWorkerClassLoader(Class<? extends WorkerDaemonProtocol> serverImplementationClass, WorkerDaemonAction<T> action, T spec, Iterable<File> classpath, Iterable<String> sharedPackages) {
+        ClassLoader workerClassLoader = createWorkerClassLoader(serverImplementationClass, classpath, sharedPackages);
+        try {
+            LOGGER.info("Executing {} in in-process worker.", action.getDescription());
+            Callable<?> worker = transferWorkerIntoWorkerClassloader(action, spec, workerClassLoader);
+            Object result = worker.call();
+            DefaultWorkResult workResult = transferResultFromWorkerClassLoader(result);
+            LOGGER.info("Successfully executed {} in in-process worker.", action.getDescription());
+            return workResult;
+        } catch (Exception e) {
+            LOGGER.info("Exception executing {} in in-process worker: {}.", action.getDescription(), e);
+            throw UncheckedException.throwAsUncheckedException(e);
+        }
+    }
+
+    private ClassLoader createWorkerClassLoader(Class<? extends WorkerDaemonProtocol> serverImplementationClass, Iterable<File> classpath, Iterable<String> sharedPackages) {
+        ClassLoader isolatedWorkerClassLoader = classLoaderFactory.createIsolatedClassLoader(new DefaultClassPath(classpath));
+        FilteringClassLoader.Spec isolatedFilteringSpec = new FilteringClassLoader.Spec();
+        for (String sharedPackage : sharedPackages) {
+            isolatedFilteringSpec.allowPackage(sharedPackage);
+        }
+        ClassLoader filteredWorkerClassLoader = classLoaderFactory.createFilteringClassLoader(isolatedWorkerClassLoader, isolatedFilteringSpec);
+        ClassLoader workerServerClassLoader = serverImplementationClass.getClassLoader();
+        return new CachingClassLoader(new MultiParentClassLoader(filteredWorkerClassLoader, workerServerClassLoader));
```

Comment:
Yeah, that's the most controversial change in this PR.

My rationale is that worker daemons and in-process workers now get the same classpath. This includes compiler workers. This change makes it so in-process workers comply with what is done in `AbstractDaemonCompiler.CompilerDaemonServer` to set up the classpath. Before the change it was simply ignored for in-process compiler workers.

@adammurdoch, could you comment on this aspect of the changes?